### PR TITLE
[3777] Support any object as semantic element in the tree representation

### LINKED ISSUE #3777: Add support for any kind of object as semantic element in the tree representation
As a developer, I need a way to associate any semantic object with a tree representation.

This work has been described in [ADR 152](https://github.com/eclipse-sirius/sirius-web/blob/master/doc/adrs/152_add_support_for_any_kind_of_object_as_semantic_element_in_the_tree_description.adoc)

### PR DESCRIPTION
**Pull request by jerome-obeo**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3777

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java [hunk 1]
```diff
@@ -66,6 +76,8 @@ public final class TreeDescription implements IRepresentationDescription {
 
     private BiFunction<VariableManager, String, IStatus> renameHandler;
 
+    private Function<VariableManager, Object> objectProvider;
```
Comment (sbegaudeau) [inline packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java]: This is not how it was expected to be named

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerDescriptionProvider.java [hunk 2]
```diff
@@ -248,7 +255,9 @@ private IStatus getDeleteHandler(VariableManager variableManager) {
     private IStatus getRenameHandler(VariableManager variableManager, String newLabel) {
         var optionalEditingContext = variableManager.get(IEditingContext.EDITING_CONTEXT, IEditingContext.class);
         var optionalTreeItem = variableManager.get(TreeItem.SELECTED_TREE_ITEM, TreeItem.class);
-        if (optionalEditingContext.isPresent() && optionalTreeItem.isPresent()) {
+        var optionalTreeDescription = variableManager.get(TreeDescription.TREE_DESCRIPTION, TreeDescription.class);
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerDescriptionProvider.java]: I don't understand why the class creating the tree description needs to be given the tree description back in order to do its job...

File: packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java [hunk 3]
```diff
@@ -32,6 +32,16 @@
...
 
+    /**
+     * The variable name used for the id.
+     */
+    public static final String ID = "id";
+
+    /**
+     * The variable name used to store a reference to a tree description.
+     */
+    public static final String TREE_DESCRIPTION = "treeDescription";
```
Comment (sbegaudeau) [inline packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java]: Why this new variable, it feels very odd that the description needs to give itself to anybody

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

**Review by sbegaudeau — COMMENTED** [review]
I would also need a test, preferably with an instance of `org.eclipse.emf.ecore.EStructuralFeature.Setting` for `org.eclipse.sirius.components.domain.Entity#superTypes` as a child of `org.eclipse.sirius.components.domain.Entity` in the explorer to have a proof that it works because right now, I see some code which delegates to other pieces of code but nothing to proves that I can display "any object as semantic element in the tree representation".

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerDescriptionProvider.java (hunk 2 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerDescriptionProvider.java]: Changed, I use the Tree instead to retrieve the TreeDescription.

File: packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java (hunk 3 above)
Comment (jerome-obeo) [reply packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java]: Changed, I use the Tree instead to retrieve the TreeDescription.

**Comment by jerome-obeo** [comment]
As an example, the Entity node has a virtual node `superTypes` that displays all super types of the entity.
The semantic object associated with this element is a `org.eclipse.emf.ecore.EStructuralFeature.Setting`.

File: packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java (hunk 1 above)
Comment (jerome-obeo) [reply packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java]: Changed

File: packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java [hunk 4]
```diff
@@ -397,4 +398,13 @@ private IStatus getDeleteHandler(VariableManager variableManager) {
     private IStatus getRenameHandler(VariableManager variableManager, String newLabel) {
         return new Failure("");
     }
+
+    private Object getObject(VariableManager variableManager) {
```
Comment (sbegaudeau) [inline packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java]: `getTreeItemObject` to be consistent with other methods

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteObjectTreeItemEventHandler.java [hunk 5]
```diff
@@ -55,16 +58,32 @@ public boolean canHandle(IEditingContext editingContext, TreeItem treeItem) {
...
+        return result;
+    }
+
+    private Optional<TreeDescription> getTreeDescription(IEditingContext editingContext, Tree tree) {
+        return this.representationDescriptionSearchService
+        .findAll(editingContext).values().stream()
+        .filter(TreeDescription.class::isInstance)
+        .map(TreeDescription.class::cast)
+        .filter(td -> Objects.equals(td.getId(), tree.getDescriptionId()))
+        .findFirst();
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteObjectTreeItemEventHandler.java]: Can be removed and replaced inline with a simple call to `IRepresentationDescriptionSearchService#findById` instead

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExpandAllTreePathProvider.java [hunk 6]
```diff
@@ -124,4 +129,25 @@ private int addAllContents(IEditingContext editingContext, String treeItemId, in
...
+        return null;
+    }
+
+    private Optional<TreeDescription> getTreeDescription(IEditingContext editingContext, Tree tree) {
+        return this.representationDescriptionSearchService
+        .findAll(editingContext).values().stream()
+        .filter(TreeDescription.class::isInstance)
+        .map(TreeDescription.class::cast)
+        .filter(td -> Objects.equals(td.getId(), tree.getDescriptionId()))
+        .findFirst();
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExpandAllTreePathProvider.java]: Can be removed and replaced inline with a simple call to `IRepresentationDescriptionSearchService#findById` instead

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerNavigationService.java [hunk 7]
```diff
@@ -86,4 +89,25 @@ private String getItemId(Object object) {
...
+        return Optional.empty();
+    }
+
+    private Optional<TreeDescription> getTreeDescription(IEditingContext editingContext, Tree tree) {
+        return this.representationDescriptionSearchService
+        .findAll(editingContext).values().stream()
+        .filter(TreeDescription.class::isInstance)
+        .map(TreeDescription.class::cast)
+        .filter(td -> Objects.equals(td.getId(), tree.getDescriptionId()))
+        .findFirst();
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerNavigationService.java]: Can be removed and replaced inline with a simple call to `IRepresentationDescriptionSearchService#findById` instead

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameDocumentTreeItemHandler.java [hunk 8]
```diff
@@ -39,34 +39,45 @@
...
+        return result;
+    }
+
+    private Optional<TreeDescription> getTreeDescription(IEditingContext editingContext, Tree tree) {
+        return this.representationDescriptionSearchService
+        .findAll(editingContext).values().stream()
+        .filter(TreeDescription.class::isInstance)
+        .map(TreeDescription.class::cast)
+        .filter(td -> Objects.equals(td.getId(), tree.getDescriptionId()))
+        .findFirst();
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameDocumentTreeItemHandler.java]: Can be removed and replaced inline with a simple call to `IRepresentationDescriptionSearchService#findById` instead

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameObjectTreeItemEventHandler.java [hunk 9]
```diff
@@ -50,19 +58,32 @@ public boolean canHandle(IEditingContext editingContext, TreeItem treeItem, Stri
...
     }
 
+    private Optional<TreeDescription> getTreeDescription(IEditingContext editingContext, Tree tree) {
+        return this.representationDescriptionSearchService
+        .findAll(editingContext).values().stream()
+        .filter(TreeDescription.class::isInstance)
+        .map(TreeDescription.class::cast)
+        .filter(td -> Objects.equals(td.getId(), tree.getDescriptionId()))
+        .findFirst();
+    }
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameObjectTreeItemEventHandler.java]: Can be removed and replaced inline with a simple call to `IRepresentationDescriptionSearchService#findById` instead

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteDocumentTreeItemEventHandler.java [hunk 10]
```diff
@@ -38,31 +40,46 @@
...
     }
+
+    private Optional<TreeDescription> getTreeDescription(IEditingContext editingContext, Tree tree) {
+        return this.representationDescriptionSearchService
+        .findAll(editingContext).values().stream()
+        .filter(TreeDescription.class::isInstance)
+        .map(TreeDescription.class::cast)
+        .filter(td -> Objects.equals(td.getId(), tree.getDescriptionId()))
+        .findFirst();
+    }
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteDocumentTreeItemEventHandler.java]: Can be removed and replaced inline with a simple call to `IRepresentationDescriptionSearchService#findById` instead

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
You've written code that was first suspected to work, now we've seen it work in action but I'm just missing one last piece of the puzzle... the test to ensure that it will keep working in the future :)

Could you create a test class `ExplorerNonSemanticElementControllerTests` near `ExplorerExpandAllControllerTests` to subscribe to the explorer on the studio project. In this test, you should use the id of the document, domain and human entity as expanded elements (In the `TreeEventInput).

It should be:

 ```
List.of(
    StudioIdentifiers.DOMAIN_DOCUMENT.toString(),
    StudioIdentifiers.DOMAIN_OBJECT.toString(),
    StudioIdentifiers.HUMAN_ENTITY_OBJECT.toString()
)
```

if I'm not mistaken and then subscribe to this tree to validate than the tree item with the label `superTypes` is visible in the tree received. Look at `ExplorerExpandAllControllerTests` to see how you can subscribe to the tree and look at `initialExpandedTreeContentConsumer` to see how you can check the tree received.

It should serve as the first example of tests for your new features. For example, when you will have the support for `getParent` in another PR, you will be able to test that you can received the proper treePath by giving the id of the setting as input. Such a test would be way simpler than the one currently in `ExplorerTreePathControllerTests` since you could just test the call to `treePathQueryRunner` and check that the result include the proper chain of id (the three one I gave you before in this comment, domain document, domain object, human entity).

Apart from that and my small comments, everything seems good for this PR (@frouene don't hesitate to merge it once my comments will have been taken into account).

File: packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java (hunk 4 above)
Comment (jerome-obeo) [reply packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteObjectTreeItemEventHandler.java (hunk 5 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteObjectTreeItemEventHandler.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExpandAllTreePathProvider.java (hunk 6 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExpandAllTreePathProvider.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerNavigationService.java (hunk 7 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerNavigationService.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameDocumentTreeItemHandler.java (hunk 8 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameDocumentTreeItemHandler.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameObjectTreeItemEventHandler.java (hunk 9 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameObjectTreeItemEventHandler.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteDocumentTreeItemEventHandler.java (hunk 10 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DeleteDocumentTreeItemEventHandler.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerChildrenProvider.java [hunk 11]
```diff
@@ -117,8 +123,14 @@ private List<Object> getDefaultChildren(VariableManager variableManager) {
                     representationMetadata.sort(Comparator.comparing(RepresentationMetadata::getLabel));
                     result.addAll(representationMetadata);
                     List<Object> contents = this.objectService.getContents(self);
+                    if (self instanceof Entity entity) {
+                        result.add(((InternalEObject) entity).eSetting(entity.eClass().getEStructuralFeature("superTypes")));
+                    }
                     result.addAll(contents);
+                } else if (self instanceof Setting setting) {
+                    result.addAll((Collection<? extends Object>) setting.get(true));
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerChildrenProvider.java]: If we have an entity then the super types setting is a child and it needs to be considered here, I'll update the PR to take it into account

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameObjectTreeItemEventHandler.java [hunk 12]
```diff
@@ -50,19 +58,29 @@ public boolean canHandle(IEditingContext editingContext, TreeItem treeItem, Stri
...
-                this.editService.editLabel(object, labelField, newLabel);
-                return new Success(ChangeKind.SEMANTIC_CHANGE, Map.of());
+    public IStatus handle(IEditingContext editingContext, TreeItem treeItem, String newLabel, Tree tree) {
+        IStatus result = new Failure("Something went wrong while handling this rename action.");
+        var optionalTreeDescription = this.getTreeDescription(editingContext, tree);
+        if (optionalTreeDescription.isPresent()) {
+            var variableManager = new VariableManager();
+            variableManager.put(IEditingContext.EDITING_CONTEXT, editingContext);
+            variableManager.put(TreeDescription.ID, treeItem.getId());
+            var object = optionalTreeDescription.get().getTreeItemObjectProvider().apply(variableManager);
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameObjectTreeItemEventHandler.java:68]: Do not use Optional.get without Optional.isPresent first

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameDocumentTreeItemHandler.java [hunk 13]
```diff
@@ -39,34 +39,42 @@
...
-            ResourceSet resourceSet = editingDomain.getResourceSet();
-            URI uri = new JSONResourceFactory().createResourceURI(treeItem.getId());
+    public IStatus handle(IEditingContext editingContext, TreeItem treeItem, String newLabel, Tree tree) {
+        IStatus result = new Failure("Something went wrong while handling this rename action.");
+        var optionalTreeDescription = this.getTreeDescription(editingContext, tree);
+        if (optionalTreeDescription.isPresent()) {
+            var variableManager = new VariableManager();
+            variableManager.put(IEditingContext.EDITING_CONTEXT, editingContext);
+            variableManager.put(TreeDescription.ID, treeItem.getId());
+            var object = optionalTreeDescription.get().getTreeItemObjectProvider().apply(variableManager);
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/RenameDocumentTreeItemHandler.java:61]: Do not use Optional.get without Optional.isPresent first

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerNavigationService.java [hunk 14]
```diff
@@ -86,4 +89,22 @@ private String getItemId(Object object) {
...
     }
+
+    private Optional<Object> getTreeItemObject(IEditingContext editingContext, String id, Tree tree) {
+        var optionalTreeDescription = this.getTreeDescription(editingContext, tree);
+
+        if (optionalTreeDescription.isPresent()) {
+            var variableManager = new VariableManager();
+            variableManager.put(IEditingContext.EDITING_CONTEXT, editingContext);
+            variableManager.put(TreeDescription.ID, id);
+            return Optional.ofNullable(optionalTreeDescription.get().getTreeItemObjectProvider().apply(variableManager));
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/ExplorerNavigationService.java:100]: Do not use Optional.get without Optional.isPresent first

**Review by sbegaudeau — COMMENTED** [review]
I'll update the PR and ensure that it gets merged

**Review by sbegaudeau — APPROVED** [review]
@jerome-obeo @frouene  With this PR, we know what we have to do to get started with the support for non-EMF elements in the explorer. I'll merge it as is today but it will be necessary to extract this hardcoded customization for Entity and Setting out of the explorer description in order to offer the same capabilities to everybody. The goal will not to add more and more hardcoded customization dedicated only to our specific test case.

**Comment by jerome-obeo** [comment]
> @jerome-obeo @frouene With this PR, we know what we have to do to get started with the support for non-EMF elements in the explorer. I'll merge it as is today but it will be necessary to extract this hardcoded customization for Entity and Setting out of the explorer description in order to offer the same capabilities to everybody. The goal will not to add more and more hardcoded customization dedicated only to our specific test case.

@sbegaudeau you are right and I already started the definition of the Domain Explorer using the new DSL for expressing Tree description (I'm currently introducing). This tree description will contain Entity and Setting customization as an example of tree customization. This description will be located in the `sirius-web-application` module.

**Pull request merged**